[KAFKA-6144]: Allow serving interactive queries from in-sync Standbys

vinothchandar opened the pull request:
- Rough initial cut, with basic HA working. Many more changes needed still
     - [TODO] Introduce a new version for AssignmentInfo & support version probing
     - [TODO] Implement API changes in KafkaStreams (deprecation, new classes, new methods)
     - [TODO] Tests using metadataForKey(), redone over new APIs. New tests with standbys involved.
     - [TODO] Tests need to be (re)written, including version probing ones

*More detailed description of your change,
if necessary. The PR title and PR message become
the squashed commit message, so use a separate
comment to ping reviewers.*

*Summary of testing strategy (including rationale)
for the feature or bug fix. Unit and/or integration
tests are expected for any behaviour change and
system tests should be considered for larger changes.*

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

vinothchandar committed:
[KAFKA-6144]: Allow serving interactive queries from in-sync Standbys

     - Rough initial cut, with basic HA working. Many more changes needed still
     - [TODO] Introduce a new version for AssignmentInfo & support version probing
     - [TODO] Implement API changes in KafkaStreams (deprecation, new classes, new methods)
     - [TODO] Tests using metadataForKey(), redone over new APIs. New tests with standbys involved.
     - [TODO] Tests need to be (re)written, including version probing ones

brary committed:
Adding KeyQueryMetadata support for reads during rebalancing

vinothchandar commented on streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java:
```diff
@@ -1104,16 +1106,43 @@ public void cleanUp() {
+     * @return {@link StreamsMetadata} for the {@code KafkaStreams} instance with the provided {@code storeName} and
      * {@code key} of this application or {@link StreamsMetadata#NOT_AVAILABLE} if Kafka Streams is (re-)initializing
      */
+    @Deprecated
```
Should we add docs that mention what alternative method to use?

vinothchandar replied:
On my earlier comment on the testing for deprecated methods.. I checked `KafkaStreams#close(long, TimeUnit)` and the deprecated `KafkaStreams` constructors and they have no tests. So lets go with what you have.

vinothchandar commented on streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java:
```diff
@@ -1104,16 +1106,43 @@ public void cleanUp() {
+    public <K> KeyQueryMetadata queryMetadataForKey(final String storeName,
+                                                    final K key,
+                                                    final Serializer<K> keySerializer) {
+        validateIsRunning();
```
Ultimately, this checks if state is RUNNING or REBALANCING? Rename to `validateIsRunningOrRebalancing` ? or to something more contextual like `validateMetadataAccessible()` and also reword the exception in `validateIsRunning()` ?

brary replied:
Exactly my thought. Renaming it to validateIsRunningOrRebalancing(). It tells more about what it happening inside IMO validateMetadataAccessible() might be confusing for some people.

vinothchandar commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/KeyQueryMetadata.java:
```diff
@@ -0,0 +1,93 @@
+import java.util.Collections;
+import java.util.Set;
+
+public class KeyQueryMetadata {
```
Some class level Java docs?

brary replied:
Done.

vinothchandar commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/KeyQueryMetadata.java:
```diff
@@ -0,0 +1,93 @@
+            return false;
+        }
+        final KeyQueryMetadata that = (KeyQueryMetadata) obj;
+        if (activeHost != that.activeHost) {
```
check using equals?  Also could be just use `Objects.equals` here?

brary replied:
Done.

vinothchandar commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/KeyQueryMetadata.java:
```diff
@@ -0,0 +1,93 @@
+    }
+
+    @Override
+    public int hashCode() {
```
same. `Objects.hashCode` won't be sufficient? may be it will break if java version < 1.7? We can leave it as-is if you are also unsure.

brary replied:
Done.

vinothchandar commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsMetadataState.java:
```diff
@@ -116,7 +116,7 @@ public String toString(final String indent) {
      * {@link DefaultStreamPartitioner} to locate the store. If a custom partitioner has been used
      * please use {@link StreamsMetadataState#getMetadataWithKey(String, Object, StreamPartitioner)}
-     *
+     * <p>
```
extra?

brary replied:
Done.

vinothchandar commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsMetadataState.java:
```diff
@@ -153,18 +154,100 @@ public String toString(final String indent) {
+     * @return The {@link KeyQueryMetadata} for the storeName and key or {@link KeyQueryMetadata#NOT_AVAILABLE}
+     * if streams is (re-)initializing
+     */
+    public synchronized <K> KeyQueryMetadata getKeyQueryMetadataWithKey(final String storeName,
```
can some of the code be shared by both these methods? by pushing common logic to another private method?

vinothchandar commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsMetadataState.java:
```diff
@@ -153,18 +154,100 @@ public String toString(final String indent) {
+            // global stores are on every node. if we dont' have the host info
+            // for this host then just pick the first metadata
+            if (thisHost == UNKNOWN_HOST) {
+                return new KeyQueryMetadata(allMetadata.get(0).hostInfo(), null, 0);
```
how about using empty lists instead of null when there is no standby? Also -1 for the case where we don't know the partition as well?

brary replied:
Done.

vinothchandar commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsMetadataState.java:
```diff
@@ -153,18 +154,100 @@ public String toString(final String indent) {
+
+    /**
+     * Find the {@link StreamsMetadata}s for a given storeName and key.
+     * <p>
```
closing `</p>` ?

vinothchandar commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsMetadataState.java:
```diff
@@ -222,31 +309,77 @@ private boolean hasPartitionsForAnyTopics(final List<String> topicNames, final S
     }
 
-    private void rebuildMetadata(final Map<HostInfo, Set<TopicPartition>> currentState) {
+    private Set<String> getStoresOnHost(final Map<String, List<String>> stores, final Set<TopicPartition> partitionsForHost) {
```
rename `stores` more descriptively?  `topicsForStore` ?

vinothchandar commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsMetadataState.java:
```diff
@@ -222,31 +309,77 @@ private boolean hasPartitionsForAnyTopics(final List<String> topicNames, final S
+            activeStoresOnHost.addAll(globalStores);
+            final Set<TopicPartition> standbypartitionsForHost = new HashSet<>();
+            final Set<String> standbyStoresOnHost = new HashSet<>();
+            if (standbyPartitionHostMap != null && standbyPartitionHostMap.containsKey(key)) {
```
empty map to indicate absence of standbys? since thats how AssignmentInfo models it?

brary replied:
Done.

vinothchandar commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsMetadataState.java:
```diff
@@ -222,31 +309,77 @@ private boolean hasPartitionsForAnyTopics(final List<String> topicNames, final S
+            matchingPartitions.add(new TopicPartition(sourceTopic, partition));
+        }
+
+        HostInfo activeHost = new HostInfo("unavailable", -1);
```
just wondering if there was an UNKNOWN partition constant somewhere with `-1` value. :) . could not find myself anyways

brary replied:
Yeah, I don't think it's there.

vinothchandar commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsMetadataState.java:
```diff
@@ -222,31 +309,77 @@ private boolean hasPartitionsForAnyTopics(final List<String> topicNames, final S
+                                                           final StreamPartitioner<? super K, ?> partitioner,
+                                                           final SourceTopicsInfo sourceTopicsInfo) {
+
+        final Integer partition = partitioner.partition(sourceTopicsInfo.topicWithMostPartitions, key, null, sourceTopicsInfo.maxPartitions);
```
it would nice to share the same code structure with `getStreamsMetadataForKey` instead of repeating. but since we plan to deprecate that in the longer run. makes sense to have it duplicated for now. (thinking out loud)

vinothchandar commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java:
```diff
@@ -617,32 +618,41 @@ public GroupAssignment assign(final Cluster metadata, final GroupSubscription gr
                     }
 
-                    partitionsByHostState.put(hostInfo, topicPartitions);
+                    // TODO: why do we have standby tasks even for non stateful sub topologies?
```
note to self: need to resolve this todo ultimately.

vinothchandar commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/AssignmentInfo.java:
```diff
@@ -150,9 +159,17 @@ public ByteBuffer encode() {
+                    out.writeInt(usedVersion);
+                    out.writeInt(commonlySupportedVersion);
+                    encodeActiveAndStandbyTaskAssignment(out);
+                    encodePartitionsByHostAsDictionary(out);
```
Can we can encode the topicname index, just once instead of duplicating once for active and standby. wdyt? I can take up this change.

brary replied:
Sure.

vinothchandar commented on streams/src/main/java/org/apache/kafka/streams/state/internals/StreamThreadStateStoreProvider.java:
```diff
@@ -46,12 +48,18 @@ public StreamThreadStateStoreProvider(final StreamThread streamThread) {
             return Collections.emptyList();
         }
-        if (!streamThread.isRunningAndNotRebalancing()) {
+        // TODO: This needs to be rethought
```
do you think this change is fine?

vinothchandar reviewed (commented):
Guess we need to add more tests in all these test files, around standby and partition number as well? overall LGTM..

vinothchandar commented:
@brary On the offset lag information, I think this information is maintained inside `StandbyTask##offsetLimits` ? We can add additional map here to track the lag based on the records seen in `StandbyTask#update(..)` .. WDYT?

vinothchandar commented:
and `StreamTask` probably contains some information on current offsets being processed.. but not sure about the current commited offset limits like for standby.. Also we need to understand how all this is updated during rebalancing as well. Take a pass. I will also do the same  and we can sync.

brary committed:
Resolving code review comments

brary committed:
Adding changes related to fetching offset lag on query

brary committed:
Reverting offset lag function to as discussed in KIP

brary committed:
Adding support to fetch checkpointed offsets and calculate lag

brary committed:
Minor checkstyle fixes and returning the planned offsetLag map

vinothchandar committed:
Minor cleanups

 - Reimplemented KafkaStreams#allLocalOffsetLags()
 - Moved KeyQueryMetadata to org.apache.kafka.streams so we can import publicly
 - Minor renames, whitespace fixes

vinothchandar committed:
Re-implement KafkaStreams#allLocalOffsetLags() by fetching endOffsets from Kafka

 - Re-implement the assignment protocol to serialize the topic index just once
 - Code cleanups, small bug fixes

vinothchandar committed:
Fix issue with reading standbyHostPartition maps for version <= 5

 - Also cleanup non-essential code formatting changes

vinothchandar committed:
Adding tests for standby streams metadata

brary committed:
Fixing shouldApplyUpdatesToStandbyStore test case

brary committed:
Addressing code review comments

brary committed:
Fixing the checkstylese

vinothchandar committed:
Changing few other maps to concurrent maps for safer iteration

vinothchandar committed:
Adding tests for KafkaStreams#allLocalOffsetLags()

vinothchandar committed:
Introduce a LagInfo class that holds both raw offset position, as well as offset lag

  - Improved javadocs for new class and KeyQueryMetadata
  - Redid the tests, based on new APIs

vinothchandar committed:
Cleanup code bases on CR comments

vinothchandar committed:
Bringing KafkaStreams#metadataForKey() back into QueryableStateIntegrationTest + few cleanups

vinothchandar committed:
Adding LagFetchIntegrationTest to test the lag values obtained during rebalancing/restoration

vinothchandar committed:
Small change to enable LagInfo#equals() for test coverage
